Libraries/VtkVgCore/vtkVgSGIReader.cxx

```diff
@@ -0,0 +1,358 @@
+  SGI_SF_RLE = 1,
+};
+
+enum SGI_ColorMap : uint32_t
+{
+  SGI_CM_Normal = 0,
+  SGI_CM_Dithered = 1,
+  SGI_CM_Screen = 2,
+  SGI_CM_Colormap = 3,
+};
+
+struct sgiImageHeader
+{
+  uint16_t          Magic;
+  SGI_StorageFormat StorageFormat;
+  uint8_t           BytesPerPixelChannel;
+  uint16_t          NumDimensions;
+  uint16_t          XDimension;
+  uint16_t          YDimension;
+  uint16_t          ZDimension;
+  uint32_t          MinValue;
+  uint32_t          MaxValue;
+  uint32_t          Dummy1; // not used
+  char              ImageName[80];
+  SGI_ColorMap      ColorMap;
+  char              Dummy2[404];
+};
+
+//-----------------------------------------------------------------------------
+void DiskToHardware(uint8_t&) {}
+void DiskToHardware(uint16_t& val) { val = ntohs(val); }
+void DiskToHardware(uint32_t& val) { val = ntohl(val); }
+
+//-----------------------------------------------------------------------------
+template <typename Enum>
+void DiskToHardware(Enum& enumValue)
+{
+  using Underlying = typename std::underlying_type<Enum>::type;
+  auto& castedValue = reinterpret_cast<Underlying&>(enumValue);
+  DiskToHardware(castedValue);
+}
+
+} // namespace <anonymous>
+
+//-----------------------------------------------------------------------------
+class vtkVgSGIReader::vtkInternal
+{
+public:
+  vtkSmartPointer<vtkImageData> ImageData;
+};
+
+//-----------------------------------------------------------------------------
+vtkVgSGIReader::vtkVgSGIReader() : Internal{new vtkInternal}
+{
+  this->SetNumberOfInputPorts(0);
+  this->SetNumberOfOutputPorts(1);
+}
+
+//-----------------------------------------------------------------------------
+vtkVgSGIReader::~vtkVgSGIReader()
+{
+  delete this->Internal;
+}
+
+//-----------------------------------------------------------------------------
+void vtkVgSGIReader::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os, indent);
+}
+
+//-----------------------------------------------------------------------------
+int vtkVgSGIReader::GetNumberOfLevels() const
+{
+  return 1;
+}
+
+//-----------------------------------------------------------------------------
+void vtkVgSGIReader::GetDimensions(int dim[2])
+{
+  if (this->Internal->ImageData)
+    {
+    int fulldim[3];
+    this->Internal->ImageData->GetDimensions(fulldim);
+    dim[0] = fulldim[0];
+    dim[1] = fulldim[1];
+    }
+  else
+    {
+    dim[0] = dim[1] = 0;
+    }
+}
+
+//----------------------------------------------------------------------------
+bool vtkVgSGIReader::CanRead(const std::string& source)
+{
+  std::string ciSource = vtksys::SystemTools::LowerCase(source);
+
+  // Check if it is a filename or if the argument is just an extension.
+  std::string ext = vtksys::SystemTools::GetFilenameLastExtension(ciSource);
+  if (ext.compare(".sgi") == 0 || (ciSource.compare("sgi") == 0) ||
+      (ciSource.compare(".sgi") == 0))
+    {
+    return true;
+    }
+
+  return false;
```

Comment:
~~Should this also check if the file exists?~~ Redacted after examining the implementation of `vtkVgPNGReader` and seeing the similarity.

Reply:
Right :slightly_smiling_face:.

The purpose of this method as I understand it is to check if the reader can read files whose *type* is described by `source`, not necessarily a *specific* file. Hence why it's set up such that `CanRead("sgi")` will return `true`.

The usual use case is to pick a representative file name (or even just the extension) and call this method on *all available readers* in order to determine which reader should be used.